File: src/store/events.ts
import { listen } from '@tauri-apps/api/event';
import type {UnlistenFn } from '@tauri-apps/api/event';
// Every variant is wrapped in `payload` because the Rust side uses
// serde(tag = "type", content = "payload") on AppEvent.
// Unit variants (no fields) are the exception — they serialize as
// just { "type": "UndoStackEmpty" } with no payload key.
export type AppEvent =
  | { type: 'ScanStarted';        payload: { path: string } }
  | { type: 'ScanProgress';       payload: { scanned: number } }
  | { type: 'ScanComplete';       payload: { total: number } }
  | { type: 'DirectoryLoaded';    payload: { path: string; total: number } }
  | { type: 'PageChanged';        payload: PageState }
  | { type: 'FileMoved';          payload: { from: string; to: string; action: MoveAction } }
  | { type: 'Undone';             payload: { path: string; action: MoveAction } }
  | { type: 'UndoStackEmpty' }
  | { type: 'StatsChanged';       payload: LibraryStats }
  | { type: 'SortChanged';        payload: { order: SortOrder } }
  | { type: 'ViewCountChanged';   payload: { view_count: number } }
  | { type: 'StaleEntryRemoved';  payload: { path: string } }
  | { type: 'LibraryEmpty' }
  | { type: 'NavigationBoundary'; payload: { kind: BoundaryKind } };

export interface PageState {
  images:        ImageInfo[];
  current_index: number;
  total:         number;
  total_pages:   number;
  current_page:  number;
  view_count:    number;
}

export interface ImageInfo {
  path:        string;
  dimensions:  [number, number] | null;
  file_size:   number | null;
  date_taken:  number | null;
  load_state:  LoadState;
}

export type LoadState =
  | { type: 'Pending' }
  | { type: 'Ready';  thumbnail: number[] }
  | { type: 'Failed'; reason: string };

export interface LibraryStats {
  remaining: number;
  selected:  number;
  rejected:  number;
}

export type MoveAction   = { type: 'Select' } | { type: 'Reject' };
export type BoundaryKind = { type: 'FirstPage' } | { type: 'LastPage' };
export type SortOrder    =
  | 'NameAsc' | 'NameDesc'
  | 'DateModifiedAsc' | 'DateModifiedDesc'
  | 'SizeAsc' | 'SizeDesc';

/// Start listening for all core events.
/// Call once at app startup. Returns an unlisten function to clean up.
export async function startCoreListener(
  onEvent: (event: AppEvent) => void
): Promise<UnlistenFn> {
  return listen<AppEvent>('core-event', (event) => {
    onEvent(event.payload);
  });
}

File: core/src/events.rs
use std::path::PathBuf;
use crate::image_cache::Image;
use crate::image_index::SortOrder;
use crate::stats::LibraryStats;

/// A snapshot of the current page, emitted whenever visible contents change.
///
/// Tauri note: add `#[derive(serde::Serialize, serde::Deserialize)]`
/// here and on all referenced types when wiring up Tauri.
#[derive(Debug, Clone)]
pub struct PageState {
    /// Images currently visible in the page window.
    pub images: Vec<Image>,
    /// Absolute index into the full image list where this page starts.
    pub current_index: usize,
    /// Total images remaining in the library (after any removals).
    pub total: usize,
    /// How many pages exist at this view_count.
    pub total_pages: usize,
    /// Which page we are on (0-based).
    pub current_page: usize,
    /// How many images are visible per page.
    pub view_count: usize,
}

/// Every state change the core can emit.
/// Consumers (CLI, Tauri) match on these — no polling required.
#[derive(Debug, Clone)]
pub enum AppEvent {
    /// A directory was scanned and the library is ready.
    /// Always followed by PageChanged (or LibraryEmpty) and StatsChanged.
    DirectoryLoaded {
        path: PathBuf,
        total: usize,
    },

    /// The visible page changed — redraw the image grid.
    /// Emitted after: load_dir, next, prev, act_on_current_at,
    /// undo, set_view_count, set_sort_order, stale cleanup.
    PageChanged(PageState),

    /// A file was successfully moved to selected/ or rejected/.
    FileMoved {
        from: PathBuf,
        to: PathBuf,
        action: MoveAction,
    },

    /// The last action was successfully undone — file is back in the library.
    Undone {
        /// The file's restored path (back in original directory).
        path: PathBuf,
        /// What action was undone.
        action: MoveAction,
    },

    /// Undo was requested but the stack was empty.
    UndoStackEmpty,

    /// Library stats changed — update counters in the sidebar.
    StatsChanged(LibraryStats),

    /// Sort order changed — the grid has been re-ordered.
    SortChanged {
        order: SortOrder,
    },

    /// View count (images per page) changed.
    ViewCountChanged {
        view_count: usize,
    },

    /// A stale index entry was silently removed (file no longer on disk).
    StaleEntryRemoved {
        path: PathBuf,
    },

    /// The library has no more images — show empty state in GUI.
    LibraryEmpty,

    /// Navigation hit a boundary — useful for disabling prev/next buttons.
    NavigationBoundary {
        kind: BoundaryKind,
    },
}

/// Whether a file was selected or rejected.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MoveAction {
    Select,
    Reject,
}

/// Which boundary was hit during navigation.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum BoundaryKind {
    /// Already on first page, prev() was a no-op.
    FirstPage,
    /// Already on last page, next() was a no-op.
    LastPage,
}